modules/livestore/partition_reader.go

```diff
@@ -0,0 +1,259 @@
+package livestore
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"strconv"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/dskit/backoff"
+	"github.com/grafana/dskit/multierror"
+	"github.com/grafana/dskit/services"
+	"github.com/grafana/tempo/pkg/ingest"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/twmb/franz-go/pkg/kadm"
+	"github.com/twmb/franz-go/pkg/kerr"
+	"github.com/twmb/franz-go/pkg/kgo"
+	"github.com/twmb/franz-go/plugin/kprom"
+)
+
+type record struct {
+	tenantID  string
+	content   []byte
+	offset    int64
+	timestamp time.Time
+}
+
+func fromKGORecord(rec *kgo.Record) record {
+	return record{
+		tenantID:  string(rec.Key),
+		content:   rec.Value,
+		offset:    rec.Offset,
+		timestamp: rec.Timestamp,
+	}
+}
+
+type consumeFn func(context.Context, []record, time.Time) error
+
+type PartitionReader struct {
+	services.Service
+
+	partitionID   int32
+	consumerGroup string
+	topic         string
+
+	client *kgo.Client
+	adm    *kadm.Client
+
+	consume consumeFn
+	metrics partitionReaderMetrics
+
+	logger log.Logger
+}
+
+func NewPartitionReaderForPusher(client *kgo.Client, partitionID int32, cfg ingest.KafkaConfig, consume consumeFn, logger log.Logger, reg prometheus.Registerer) (*PartitionReader, error) {
+	metrics := newPartitionReaderMetrics(partitionID, reg)
+	return newPartitionReader(client, partitionID, cfg, consume, logger, metrics)
+}
+
+func newPartitionReader(client *kgo.Client, partitionID int32, cfg ingest.KafkaConfig, consume consumeFn, logger log.Logger, metrics partitionReaderMetrics) (*PartitionReader, error) {
+	r := &PartitionReader{
+		partitionID:   partitionID,
+		consumerGroup: cfg.ConsumerGroup,
+		topic:         cfg.Topic,
+		client:        client,
+		adm:           kadm.NewClient(client),
+		consume:       consume,
+		metrics:       metrics,
+		logger:        log.With(logger, "partition", partitionID),
+	}
+
+	r.Service = services.NewBasicService(r.start, r.running, r.stop)
+	return r, nil
+}
+
+func (r *PartitionReader) start(context.Context) error {
+	return nil
+}
+
+func (r *PartitionReader) running(ctx context.Context) error {
+	offset, err := r.fetchLastCommittedOffsetWithRetries(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to fetch last committed offset: %w", err)
+	}
+
+	r.client.AddConsumePartitions(map[string]map[int32]kgo.Offset{r.topic: {r.partitionID: offset}})
+	defer r.client.RemoveConsumePartitions(map[string][]int32{r.topic: {r.partitionID}})
+
+	for ctx.Err() == nil {
+		fetches := r.client.PollFetches(ctx)
```

Comment:
This loop as written will lead to excessive commits because it is committing for every read.  A synchronous loop is still ok, but it needs to probably check against the current time and try to target a commit every couple seconds or something.   Also, need a timeout on PollFetches because by default it waits forever.  In the case that we've consumed the last message in the queue, it would hang and not be able to commit until something else showed up.